### PR TITLE
Add Clock::sleep_until method

### PR DESCRIPTION
**Pull request by emersonknapp**
Part of #1730 
Related to ros2/rosbag2#694

Add a `Clock::sleep_until` method. Handle all 3 clock types, and edge cases for shutdown and enabling/disabling ROS time.

Adds test cases for enable/disable ROStime, and `rclcpp::shutdown` - which needed special interrupts. Tests ROS time via `/clock` messages, and has basic testing for steady/system times, though those are much simpler cases.

### CODE REVIEW

**Comment by emersonknapp** [comment]
For curiosity while awaiting feedback:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15046/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9781/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12734/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15213/)

**Review by fujitatomoya — COMMENTED** [review]
overall, i think looks good. some minor comments.

File: rclcpp/src/rclcpp/clock.cpp
Comment (emersonknapp) [reply rclcpp/src/rclcpp/clock.cpp:129]: No, we need this jump handler to detect when `ros_time_is_active` changes. The `on_clock_change = true` makes it so that the callback is triggered when ros_time changes in either direction. So, if it was really system time, but ros time becomes active, we want to quit sleeping because system time is no longer relevant.

**Comment by emersonknapp** [comment]
@fujitatomoya do you have merge access? Or do we also need reviews from one of the maintainers @mabelzhang @wjwwood @ivanpauno

**Comment by emersonknapp** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15210/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9930/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12890/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15392/)

**Comment by fujitatomoya** [comment]
> do we also need reviews from one of the maintainers

👍 we could use the other review. @ivanpauno @clalancette @wjwwood

**Review by fujitatomoya — APPROVED** [review]

File: rclcpp/test/rclcpp/test_time.cpp
Comment (clalancette) [inline rclcpp/test/rclcpp/test_time.cpp:523]: This feels like it isn't going to be long enough to not be flaky, particularly on Windows.  I would suggest we set this much higher, maybe something like 200 milliseconds.

File: rclcpp/src/rclcpp/clock.cpp
Comment (clalancette) [inline rclcpp/src/rclcpp/clock.cpp:102]: Can we add a test for this situation?

File: rclcpp/src/rclcpp/clock.cpp
Comment (clalancette) [inline rclcpp/src/rclcpp/clock.cpp:169]: I think it would be nice to have a test for when this is *not* the case, i.e. when the user passed an `until` that was in the past.

**Review by clalancette — CHANGES_REQUESTED** [review]
A few minor things that I think we should fix, but otherwise looks pretty good to me.

File: rclcpp/src/rclcpp/clock.cpp
Comment (emersonknapp) [reply rclcpp/src/rclcpp/clock.cpp:102]: Added

File: rclcpp/src/rclcpp/clock.cpp
Comment (emersonknapp) [reply rclcpp/src/rclcpp/clock.cpp:169]: Added

File: rclcpp/include/rclcpp/clock.hpp
Comment (clalancette) [inline rclcpp/include/rclcpp/clock.hpp:91]: I forgot this on the last review, but it might be nice to mention what happens when `until` is in the past (i.e. it returns true immediately).

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left one nit inline, but it just changing a documentation string.

More concerning is the yellow builds on aarch64; it looks like one of these tests is probably crashing there.  Once that is fixed, I'm happy to approve.

**Comment by emersonknapp** [comment]
little sanity check aarch64 [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/10004/)

**Comment by emersonknapp** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15293/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10008/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12974/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15485/)

**Review by clalancette — APPROVED** [review]

**Comment by emersonknapp** [comment]
@clalancette are you able to merge this, based on that approval?

**Comment by clalancette** [comment]
Yeah, I'll go ahead and merge.

**Comment by clalancette** [comment]
This is flakey on aarch64: https://ci.ros2.org/job/ci_linux-aarch64/10018/

**Comment by emersonknapp** [comment]
Yeah that's weird. I'm looking into it, can spin up an ARM instance on EC2

**Pull request merged**
